--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -812,6 +812,17 @@ static struct board info __initdata board_96333aw2g = {
 	},
 };
 
+static struct board_info __initdata board_96333w = {
+	.name = "96333W",
+	.expected_cpu_id = 0x6338,
+
+	.has_enet0 = 1,
+	.enet0 = {
+		.has_phy = 1,
+		.use_internal_phy = 1,
+	},
+};
+
 static struct board_info __initdata board_96338gw = {
 	.name				= "96338GW",
 	.expected_cpu_id		= 0x6338,
@@ -2738,6 +2749,7 @@ static const struct board_info __initconst *bcm963xx_boards[] = {
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6338
 	&board_96333aw2g,
+	&board_96333w,
 	&board_96338gw,
 	&board_96338w,
 	&board_96338w2_e7t,
@@ -2847,6 +2859,7 @@ static struct of_device_id const bcm963xx_boards_dt[] = {
 #endif
 #ifdef CONFIG_BCM63XX_CPU_6338
 	{ .compatible = "brcm,bcm96333aw2g", .data = &board_96333aw2g, },
+	{ .compatible = "brcm,bcm96333w", .data = &board_96333w, },
 	{ .compatible = "brcm,bcm96338gw", .data = &board_96338gw, },
 	{ .compatible = "brcm,bcm96338w", .data = &board_96338w, },
 	{ .compatible = "dynalink,rta1320", .data = &board_rta1320_16m, },
